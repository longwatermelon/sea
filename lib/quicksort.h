fn quicksort(arr:int*, n:int) -> void;
