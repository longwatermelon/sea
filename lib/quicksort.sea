fn swap(x:int*, y:int*) -> void {
    let t:int = *x;
    *x = *y;
    *y = t;
}

fn qsort_rec(arr:int*, lo:int, hi:int) -> void {
    if (lo>=hi) { return 0; }

    let i:int = lo;
    let j:int = hi;
    let mid:int = lo + (hi - lo)/2;
    let pivot:int = arr[mid];

    while (i<=j) {
        while (arr[i] < pivot) { i = i + 1; }
        while (arr[j] > pivot) { j = j - 1; }
        if (i<=j) {
            swap(arr + i, arr + j);
            i = i + 1;
            j = j - 1;
        }
    }

    if (lo < j) { qsort_rec(arr, lo, j); }
    if (i < hi) { qsort_rec(arr, i, hi); }
}

fn quicksort(arr:int*, n:int) -> void {
    if (n<=1) { return 0; }
    qsort_rec(arr, 0, n-1);
}

