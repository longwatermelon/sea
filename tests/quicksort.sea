#include "../lib/quicksort.h"

fn main() -> int {
    let n:int = 6;
    let a:int* = stalloc(6, int);
    a[0]=5; a[1]=1; a[2]=4; a[3]=2; a[4]=8; a[5]=0;
    quicksort(a, n);

    let i:int = 0;
    let sum:int = 0;
    while (i<n) {
        sum = sum + a[i]*(i+1);
        i = i + 1;
    }
    return sum;
}

