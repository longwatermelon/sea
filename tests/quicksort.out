RETURN_CODE:97
STDOUT:

